### DISCUSSION TITLE
Support of Sunny Boy Storage 3.7

### DISCUSSION THREAD

**Discussion by dh2wq** [SBFspot/SBFspot#518] [General]
Hi,
many thanks for this great tool. For me, it works great for Power Values, but I don't get any information regarding the battery. 
I get following information from the inverter?
```
Device Class:     Battery Inverter
Device Type:      Sunny Boy Storage 3.7
Software Version: 03.12.33.R
```
But for Battery Charge Status or general Battery Info I get just the Packet Status 21 which I understand means "command not supported". 
```
Packet status: 21
getBatteryChargeStatus returned an error: 21
Packet status: 21
getBatteryInfo returned an error: 21
```
I checked the addresses of the return values in the diff description of the sunny portal, and they seem to be the same, but the command may be different. Are the commands somewhere available? I haven't found on the web so far. 

I also get odd response for getSpotDCPower and getSpotDCVoltage (I dont see if this comes from Sunny Boy Storage 3.7 or Sunny Tripower 10.0)
```
Packet status: 21
getSpotDCPower returned an error: 21
Packet status: 21
getSpotDCVoltage returned an error: 21
00464B01 28 Sun Jan 30 19:45:38 2022 'Grid voltage phase L1 against L2' NaN NaN NaN NaN
00464C01 28 Sun Jan 30 19:45:38 2022 'Grid voltage phase L2 against L3' NaN NaN NaN NaN
00464D01 28 Sun Jan 30 19:45:38 2022 'Grid voltage phase L3 against L1' NaN NaN NaN NaN
00464E01 28 Sun Jan 30 19:45:52 2022 '' 95 95 95 95
00464B07 28 Sun Jan 30 19:45:52 2022 'Grid voltage phase L1 against L2' NaN NaN NaN NaN
00464C07 28 Sun Jan 30 19:45:52 2022 'Grid voltage phase L2 against L3' NaN NaN NaN NaN
00464D07 28 Sun Jan 30 19:45:52 2022 'Grid voltage phase L3 against L1' NaN NaN NaN NaN
00464E07 28 Sun Jan 30 19:45:52 2022 '' 95 95 95 95

```
I do have basic skills in programming and may support here, but I am not keen enough to just loop commands to the Inverter for now. I am scared I screw something up.

BR
Olli

**Comment by sbf-** [comment]
So you have an SB Storage and an STP 10, connected via ethernet I suppose
When using -v5 you will get more verbose output.
For the battery info issue I'll need a debug log (-d5 -v5) In case you don't want to post it here, you can send it to 'sbfspot at proximus dot be'

**Reply by dh2wq** [reply]
I guess I find the issue and it is easy to solve for the battery case, but I'm afraid this might be an issue for other commands as well.

It looks like if one Device in the list doesn't support a command which one of the others support, this isn't handled correctly. 

To make it easier to understand, please look at this code:
If any of the inverters in `m_inverters` doesn't support `BatteryChargeStatus` `getInverterData()` will return `21` and the else case isn't handled.
https://github.com/SBFspot/SBFspot/blob/e6aafe2a312f18f9054671902621adf3850d7fc8/SBFspot/Inverter.cpp#L174-L176

As a hotfix, I implemented a hack to handle this for the Battery Information command:
```
        {
            if ((devList[i]->hasBattery == false) && (command == 0x51000200))
                return E_OK;
            else
                ethSend(pcktBuf, devList[i]->IPAddress);
        }
```
instead of:
https://github.com/SBFspot/SBFspot/blob/e6aafe2a312f18f9054671902621adf3850d7fc8/SBFspot/SBFspot.cpp#L2412-L2415

I haven't contributed to community projects for a while, If i should commit and push or generate a branch, please tell me.

**Comment by deagel** [comment]
Hi, have the same issue, an update sounds nice for me :)
